### PR TITLE
[AOTI] Auto-tune Triton kernels in a seperate block

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129057

Summary: Currently AOTI does a two-pass compilation for the CUDA backend. In the first pass AOTI generates Python code, runs the generated code once with real example inputs to trigger Triton kernel compilation and tuning, and then AOTI runs the second pass to generate cpp code and compiles that into a shared library.

There are several problems with this approach when we want to enable the cpp wrapper mode for JIT Inductor:
* Compilation time: JIT compilation is more sensitive to compilation time than AOT compilation. The two-pass approach does add extra overhead for compilation.
* Peak memory size: when executing the first-pass generated code with real inputs, some inputs need to be cloned to avoid side effect coming from input mutation. This can raise the high-water mark for memory consumption.
* Missing triton kernel autotuning: Because kernel autotune depends on the kernel being executed in the two-pass approach, some kernels will not be autotuned when a model contains control flow such as torch.if or torch.while.

This PR is the first step towards solving these problems by moving Triton kernel autotuning to the compile time and use random inputs for tuning. The cpp wrapper codegen still has two passes, but in the first pass, Inductor will generate a separate code just for kernel autotuning, with https://gist.github.com/desertfire/606dc772b3e989b5e2edc66d76593070 as an example, and we no longer need to execute the model after the first-pass finishes. After that we rerun a second pass to generate cpp code. This reduces peak memory consumption and enables kernel autotuning when there is control flow. Truly making the codegen into one-pass will come later once this solution is proven stable and generates as performant kernels as before.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @chauhang

Differential Revision: [D58782766](https://our.internmc.facebook.com/intern/diff/D58782766)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58782766).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58782766).

**Review by jansel — APPROVED** [review]

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58782766).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58782766).

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -162,30 +168,47 @@ def _convert_to_sympy_expr(item: Union[int, sympy.Expr]) -> sympy.Expr:
     def determine_grid(grid: TritonGrid):
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:168]: Can you add docstring & type annotations on what is being returned now ?

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -787,6 +855,33 @@ def generate(self, is_inference):
 
         return result.getvaluewithlinemap()
 
+    def generate_autotune_block(self):
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:858]: nit: `generate_and_run` - we're not just generating it, we're also running it.

File: torch/_inductor/codegen/wrapper.py [hunk 3]
```diff
@@ -1379,6 +1477,56 @@ def wrap_arg(arg):
 
         return device_index, call_args
 
+    def generate_example_arg_value(self, arg, arg_type=None, raw_arg=None, index=None):
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:1480]: We do not need to call `torch.rand` for outputs - can just can just call empty_strided

File: torch/_inductor/codegen/wrapper.py [hunk 4]
```diff
@@ -1415,6 +1564,58 @@ def generate_kernel_call(
                 self.writeline(
                     f"{name}.run({call_args_str}, grid={grid_str}, stream={stream_name})"
                 )
+                if config.triton.autotune_at_compile_time:
+                    # Create example args for autotune in a separate epilogue
+                    assert arg_types is not None and len(call_args) == len(
+                        arg_types
+                    ), "call_args and arg_types do not match"
+
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:1572]: nit: maybe could be factored out

**Review by eellison — APPROVED** [review]
Do you have any numbers on mem improvement, just curious ?

where doable would be great to have a few more type annotations and docstrings

**Comment by desertfire** [comment]
> Do you have any numbers on mem improvement, just curious ?
> 
Yes, [here](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2013%20Jun%202024%2022%3A29%3A58%20GMT&stopTime=Thu%2C%2020%20Jun%202024%2022%3A29%3A58%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&lBranch=gh/desertfire/408/head&lCommit=0d3edddad288928f7f049d28c94f8fb65d2ec681&rBranch=main&rCommit=ce79b09415c0513800d647ed516fe38d38140bfb). Torchbench on CI is down, otherwise I would expect to see a larger improvement to TB.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58782766).

**Comment by sifengyang** [comment]
@desertfire Cite the above information:
```
Peak memory size: when executing the first-pass generated code with real inputs, some inputs need to be cloned to avoid side effect coming from input mutation. This can raise the high-water mark for memory consumption.
```

In my training task, GPU memory usage was near its peak. With torch._inductor.config.cpp_wrapper=1 and torch._inductor.config.triton.autotune_at_compile_time=False enabled, the cpp_wrapper  does two-pass compilation, resulting in an OOM error.

Is there a plan to address this problem?

**Comment by desertfire** [comment]
No plan to address that in any near future. Does `autotune_at_compile_time=True` work for your use case?

**Pull request closed without merging**